### PR TITLE
[mlir][llvm] Add experimental.vector.interleave2 intrinsic

### PR DESCRIPTION
**Pull request by c-rhodes**

### CODE REVIEW

**Comment by MacDue** [comment]
Maybe add zip2 as well? 
(In the 2/4-way lowerings I think `zip1, zip2`, would use half the registers two `zip1`s, which might be something we'd like to try).

**Comment by dcaballe** [comment]
I've been successfully using `vector.shuffle` to model this pattern for Neon and the zip instructions were generated accordingly so I'm wondering if we could do the same for scalable. What is the way to model scalable shuffles in LLVM?

**Comment by banach-space** [comment]
> I've been successfully using `vector.shuffle` to model this pattern for Neon and the zip instructions were generated accordingly so I'm wondering if we could do the same for scalable. What is the way to model scalable shuffles in LLVM?

IIUC, that's not possible for SVE. From https://llvm.org/docs/LangRef.html#id189:

> For scalable vectors, the only valid mask values at present are zeroinitializer, undef and poison, since we cannot write all indices as literals for a vector with a length unknown at compile time.

**Comment by MacDue** [comment]
> IIUC, that's not possible for SVE. From https://llvm.org/docs/LangRef.html#id189:

For SVE/scalable vectors there is `@llvm.experimental.vector.interleave2` which a slightly higher level abstraction for this:
https://llvm.org/docs/LangRef.html#llvm-experimental-vector-interleave2-intrinsic

**Comment by c-rhodes** [comment]
> Maybe add zip2 as well? (In the 2/4-way lowerings I think `zip1, zip2`, would use half the registers two `zip1`s, which might be something we'd like to try).

I tried that already but couldn't get it to work

**Comment by c-rhodes** [comment]
> > Maybe add zip2 as well? (In the 2/4-way lowerings I think `zip1, zip2`, would use half the registers two `zip1`s, which might be something we'd like to try).
> 
> I tried that already but couldn't get it to work

If we can get that to work could be a nice future improvement, I'd rather not add an intrinsic until we're sure if that's ok, and it's not necessary for this first widening outer product support

**Comment by c-rhodes** [comment]
> > IIUC, that's not possible for SVE. From https://llvm.org/docs/LangRef.html#id189:
> 
> For SVE/scalable vectors there is `@llvm.experimental.vector.interleave2` which a slightly higher level abstraction for this: https://llvm.org/docs/LangRef.html#llvm-experimental-vector-interleave2-intrinsic

Oh cool, this is nicer, I'll use this intrinsic instead.

> Though the MLIR `vector.shuffle` operation currently cannot model a scalable zip, so in MLIR this would probably be `vector.scalable.interleave` (following from `vector.scalable.insert/extract`, which are also experimental LLVM intrinsics).

Sounds good 👍 I'll update the patches to use the target-agnostic interleave intrinsic, and we can look to add an op later.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td [hunk 1]
```diff
@@ -969,6 +969,20 @@ def LLVM_vector_extract
   }];
 }
 
+class HalfElementsVectorTypeConstraint<string lhs, string rhs>
+    : TypesMatchWith<
+        "'" # rhs # "'" # " has half elements of '" # lhs # "'",
+      lhs, rhs,
+      "VectorType(VectorType::Builder(::llvm::cast<VectorType>($_self))"
+        ".setDim(0, ::llvm::cast<VectorType>($_self).getShape()[0] / 2))">;
```
Comment (MacDue) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td:977]: This would also need to check the number of elements are even, right?

**Review by banach-space — APPROVED** [review]
LGTM, thanks!

File: mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td (hunk 1 above)
Comment (c-rhodes) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td]: I've simplified it by reversing the check.

**Review by dcaballe — APPROVED** [review]
Cool!

**Comment by c-rhodes** [comment]
@MacDue pointed out the input type`LLVM_AnyVector` includes `LLVMFixedVectorType` and `LLVMScalableVectorType` and `cast<VectorType>` will crash. I've updated the constraint to fix this.

**Review by MacDue — APPROVED** [review]
LGTM, thanks

**Pull request merged**